Initial quota support.

geeknoid opened the pull request:
- Introduce a simple quota aspect manager. This is largely a placeholder 
until we rejigger things next week to streamline the config story.

- Introduce a simple fully memory-resident quota adapter.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/mixer/202)
<!-- Reviewable:end -->

douglas-reid commented on adapter/memQuota/memQuota_test.go:
```diff
@@ -0,0 +1,286 @@
+		}
+
+		if amount != c.allocResult {
+			t.Errorf("Expecting %d, got %d, case %d", c.allocResult, amount, i)
```
how about `("Alloc(%v) => %d, want %d", amount, c.allocResult)` ?

matches expected order from: https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

same thing for the below messages.

geeknoid replied:
Done

douglas-reid commented on adapter/memQuota/memQuota_test.go:
```diff
@@ -0,0 +1,286 @@
+			t.Errorf("Expecting %d, got %d, case %d", c.allocResult, amount, i)
+		}
+
+		qa = adapter.QuotaArgs{
```
can we split this out into a separate TestMemQuota_Release() or similar?

geeknoid replied:
I find having it all in one place is more convenient, avoiding a lot of redundancy.

douglas-reid commented on adapter/memQuota/memQuota_test.go:
```diff
@@ -0,0 +1,286 @@
+	qa.DeduplicationID = "0"
+	amount, _ := asp.Alloc(qa)
+	if amount != 10 {
+		t.Errorf("Expecting 10, got %d", amount)
```
as above, suggest something like: `"Alloc() => %d, want 10"` here and below.

geeknoid replied:
Done

douglas-reid commented on adapter/memQuota/rollingWindow.go:
```diff
@@ -0,0 +1,85 @@
+
+package memQuota
+
+import (
```
can probably be collapsed to just `import "sync"`

geeknoid replied:
Done

douglas-reid commented on adapter/memQuota/rollingWindow.go:
```diff
@@ -0,0 +1,85 @@
+	ticksInWindow int32
+}
+
+func newRollingWindow(limit int64, ticksInWindow int32) *rollingWindow {
```
nit: just `ticks`. inWindow is implied by method name.

geeknoid replied:
Just 'ticks' doesn't do it for me, I think it's not obvious enough.

douglas-reid commented on adapter/memQuota/rollingWindow.go:
```diff
@@ -0,0 +1,85 @@
+
+func (w *rollingWindow) alloc(amount int64, currentTick int32) bool {
+	// reap old entries
+	for n := w.first; n != nil && n.expirationTick <= currentTick; n = n.next {
```
any reason not to use "container/list" (or similar) for the window/allocs here (as opposed to having the allocs maintain the pointers themselves)?

douglas-reid commented on adapter/memQuota/rollingWindow.go:
```diff
@@ -0,0 +1,85 @@
+	"sync"
+)
+
+// A single allocation, carrying its amount and expiration time
```
did we look at https://godoc.org/golang.org/x/time/rate for rate limiting? is there something our implementation provides that is better than the existing rate limiter?

geeknoid replied:
Yes, I looked at a few rate limiters. They seem to all suffer from the '11th hour' problem. That is, if you have a QPS limit of 10, it's possible for someone to allocate 10 units at 0.9s and another 10 units at 1.1s.

My approach is a true rolling window, allowing no more than 10 units in any 1 second interval.

On top of that, my version is considerably more efficient in memory & time :-)

geeknoid commented on adapter/memQuota/config/config.proto:
```diff
@@ -0,0 +1,24 @@
+
+message Params {
+	// Minimum number of seconds that deduplication is possible for a given operation.
+	int32 min_deduplication_window = 1;
```
Done

geeknoid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,309 @@
+
+	mq.Lock()
+
+	if d.Kind == adapter.AllocationKind {
```
Done

geeknoid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,309 @@
+	mq.dedupLock.Lock()
+	t := mq.oldDedup
+	mq.oldDedup = mq.recentDedup
+	mq.recentDedup = t
```
Trying to save on garbage.

geeknoid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,309 @@
+		buffer.WriteString(k)
+		buffer.WriteString("=")
+
+		// isn't it amazing that the primitive types like int64 and bool don't implement the Stringer interface?
```
Done.

geeknoid commented on adapter/memQuota/memQuota_test.go:
```diff
@@ -0,0 +1,286 @@
+		}
+
+		if err != nil {
+			t.Errorf("Expecting success, got %v, case %d", err, i)
```
Done.

Although I did do these test output changes, I find this somewhat busy work and not particularly useful to anyone. When you have a test failure, the exact line of the failure is printed. The first thing you do when you see such a failure is load the test file and go to that line. And on top of that, the fact it's highly likely that the majority of these test failures will never ever be seen.

A large test corpus I worked with previous used a framework that encourage minimal failure messages. The parameters involved in the test case would be printed in a structured way, but no additional prose was included. It made writing tests faster and was just as diagnosable.

Anyway, that's my perspective on these...

geeknoid commented:
I've addressed most issues, fixed a race condition with how I was handling dedup lookup, and substantial reduced memory churn and steady-state use.

PTAL.

mandarjog commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,335 @@
+	return result, nil
+}
+
+func (mq *memQuota) allocCore(d *adapter.QuotaDefinition, key string, amount int64, bestEffort bool) int64 {
```
can we split this into `rateQuota()` and `allocationQuota()`

mandarjog commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,305 @@
+var keyWorkspacePool = sync.Pool{New: func() interface{} { return &keyWorkspace{} }}
+
+var (
+	name = "istio/memQuota"
```
I have 2 issues with combining these 2 functions into 1.
1. ratelimit and quota are different aspect `kinds` from a user's perspective.
so 
```
r.RegisterQuota(newQuotaBuilder())
r.RegisterRateLimit(newRateLimitBuilder())
```

2. I think there is very little common between AllocationQuota and RateLimit.
Allocation() is different (windowed, and non-windowed) and release is different. 

If there *is* some common code it can be made into a base class or helper funcs.

geeknoid replied:
Why should these be different aspects? This is adding cognitive load for no value.

The only distinction between an allocation quota and a rate quota is that rate quotas expire after a while. Everything else is identical. There is in fact only a single QuotaDescriptor for this reason, we don't have different types.

geeknoid commented:
Applied changes corresponding to the update to QuotaDescriptor I've submitted to istio/api.

Basically, the line between rate & allocation quota is removed completely from the API surface. We maintain a distinction in the implementation of the adapter strictly for the sake memory efficiency for non-expiring quotas.

PTAL

douglas-reid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,296 @@
+
+			currentTick := mq.getTick()
+			if !window.alloc(result, currentTick) {
+				if bestEffort {
```
what does it mean to be "bestEffort" for a rate limit?  Is this the equivalent of a "burst" rate above and beyond the normal limit? Should we be modeling it in that manner?

I'm not clear what the expectations are for: "allow 10 qps, best effort".  

With this code, won't it be possible to send unlimited QPS, as window.available() == 0 is not a blocker?

It feels like, for rate limiting, we should be establishing a limit with a configurable burst.

geeknoid replied:
This code doesn't return errors when it can't allocate quota; it returns 0 instead. Errors are used only for actual error conditions such as misconfiguration.

Regarding the semantics, it's the semantics are:

- I can support N units per seconds
- Please allocate M units right now, or as much as you can

I think this is a fairly straightforward model. It'll work well for prefetching tokens in particular.

geeknoid replied:
Regarding bursting, this is done through a simple composition of limits. For example, a QPS limit of 100 and a QPM limit of 1000. This allows bursts of 100QPS in the broader context of 1000QPM.

douglas-reid replied:
What handles enforcement within Istio?  Is the alloc bit not responsible for ensuring rate limits are respected?

If it is, best effort, as I read the code, is the same as having no limit defined.

douglas-reid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,296 @@
+
+				mq.cells[key] = inUse - result
+			}
+		} else {
```
can we rewrite this as:

```
if d.ExpirationSeconds > 0 {
   return 0
}
....
```

(dropping the `else` and pulling the no-op condition up front).

geeknoid replied:
The condition is no longer a no-op, but I did reorg the code as suggested.

douglas-reid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,296 @@
+
+type quotaFunc func(d *adapter.QuotaDefinition, key string) int64
+
+func (mq *memQuota) common(args adapter.QuotaArgs, qf quotaFunc) (int64, error) {
```
is there a more descriptive name than `common`?

geeknoid replied:
Done

douglas-reid commented on adapter/memQuota/rollingWindow.go:
```diff
@@ -0,0 +1,71 @@
+// Implements a rolling window that allows N units to be allocated per rolling time interval.
+// Time is abstracted in terms of ticks, provided by the caller, decoupling the
+// implementation from real-time, enabling much easier testing and more flexibility.
+type rollingWindow struct {
```
at the risk of annoying repetition, I still feel like we should be using something like: https://godoc.org/golang.org/x/time/rate#Limiter here. It uses the token bucket algorithm and should avoid the 11th hour issue you've raised earlier.

Doing so has several advantages (well-tested, widely-used, part of the official supplementary libs, provides a configurable burst). It would also be easier to read and understand (defined in terms that match the user config, no "ticks" to mentally map).

The one thing it doesn't provide is insight into remaining available tokens.  But I'm not clear how that will be used by rate limiting clients in practice. This relates to my other question about what best-effort means for rate limits.

geeknoid replied:
We're talking about 1 26 line function. What's there now is tested, works, has all the semantics we need, and is very efficient. I'd like to leave as-is.

douglas-reid replied:
OK. I don't think this is a blocker.

If we are having the aspect manager do enforcement, then this seems reasonable (as the existing libraries don't return numTokens consumed, etc.).

I do worry, however, in general about reimplementing vetted solutions for problems.

douglas-reid commented on pkg/adapter/quota.go:
```diff
@@ -34,11 +34,26 @@ type (
+		MaxAmount int64
+
+		// Size of rolling window in seconds
+		ExpirationSeconds int32
```
any reason not to take a generic time.Duration here?  Do we care about flexibility for sub-second windows?

geeknoid replied:
Done

geeknoid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,296 @@
+)
+
+const (
+	ticksPerSecond = 10
```
Done

geeknoid commented on adapter/memQuota/memQuota.go:
```diff
@@ -0,0 +1,296 @@
+			if result > 0 {
+				mq.cells[key] = inUse + result
+			}
+		} else {
```
Done

geeknoid commented:
Addressed the comments, except for the suggestion to use the existing rate limiter, for reasons stated in my comments. I've also gone ahead and implement rate limit release semantics.

PTAL.

geeknoid commented:
The aspect manager is where this decision happens.

   - Alloc returns requestedAmount or 0, depending on if totalAvailable >=
   requestedAmount
   - AllocBestEffort returns requestedAmount or totalAvailable, whichever
   is least

The aspect manager will implement the allow/disallow semantics
appropriately. The proxy will depend on Alloc semantics initially, and
leverage AllocBestEffort once we integrate prefetching.


On Tue, Jan 31, 2017 at 9:41 AM, Douglas Reid <notifications@github.com>
wrote:

> *@douglas-reid* commented on this pull request.
> ------------------------------
>
> In adapter/memQuota/memQuota.go <https://github.com/istio/mixer/pull/202>:
>
> > +				}
> +			}
> +
> +			if result > 0 {
> +				mq.cells[key] = inUse + result
> +			}
> +		} else {
> +			window, ok := mq.windows[key]
> +			if !ok {
> +				window = newRollingWindow(d.MaxAmount, d.ExpirationSeconds*ticksPerSecond)
> +				mq.windows[key] = window
> +			}
> +
> +			currentTick := mq.getTick()
> +			if !window.alloc(result, currentTick) {
> +				if bestEffort {
>
> What handles enforcement within Istio? Is the alloc bit not responsible
> for ensuring rate limits are respected?
>
> If it is, best effort, as I read the code, is the same as having no limit
> defined.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/mixer/pull/202>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHQ388urcKiyazEv7J7fHKYMT1bTwks5rX3JRgaJpZM4LvczM>
> .
>

douglas-reid commented on pkg/aspect/quotaManager.go:
```diff
@@ -0,0 +1,117 @@
+			}
+		}
+
+		// TODO: handle return value!
```
just for clarity: this TODO is the bit we expect to do enforcement (based on your earlier comments), correct?

So, for this to work, this needs to be impl'd?

geeknoid replied:
Righto, I'll plug this in as I wire up Check and Quota to do something real wrt quotas.

geeknoid commented:
Added a few more tests and fixed a bug as a result. Coverage of rollingWindow.go is 100%, coverage of memQuota.go is missing 2 lines which is around the background reaper goroutine. The reaper itself is tested, just not the tick-based for loop.

Anyway, all ready for checkin :-)
